networkx/algorithms/graph_hashing.py

```diff
@@ -78,74 +100,201 @@ def weisfeiler_lehman_graph_hash(
        Graph Kernels. Journal of Machine Learning Research. 2011.
        http://www.jmlr.org/papers/volume12/shervashidze11a/shervashidze11a.pdf
-    """
 
-    def neighborhood_aggregate(G, node, node_labels, edge_attr=None):
-        """
-        Compute new labels for given node by aggregating
-        the labels of each node's neighbors.
-        """
-        label_list = [node_labels[node]]
-        for nei in G.neighbors(node):
-            prefix = "" if not edge_attr else G[node][nei][edge_attr]
-            label_list.append(prefix + node_labels[nei])
-        return "".join(sorted(label_list))
+    See also
+    --------
+    weisfeiler_lehman_subgraph_hashes
+    """
 
-    def weisfeiler_lehman_step(G, labels, edge_attr=None, node_attr=None):
+    def weisfeiler_lehman_step(G, labels, edge_attr=None):
         """
         Apply neighborhood aggregation to each node
         in the graph.
         Computes a dictionary with labels for each node.
         """
-        new_labels = dict()
+        new_labels = {}
         for node in G.nodes():
-            new_labels[node] = neighborhood_aggregate(
-                G, node, labels, edge_attr=edge_attr
-            )
+            label = _neighborhood_aggregate(G, node, labels, edge_attr=edge_attr)
+            new_labels[node] = _hash_label(label, digest_size)
         return new_labels
 
-    items = []
-    node_labels = dict()
     # set initial node labels
-    for node in G.nodes():
-        if (not node_attr) and (not edge_attr):
-            node_labels[node] = str(G.degree(node))
-        elif node_attr:
-            node_labels[node] = str(G.nodes[node][node_attr])
-        else:
-            node_labels[node] = ""
-
-    for k in range(iterations):
+    node_labels = _init_node_labels(G, edge_attr, node_attr)
+
+    subgraph_hash_counts = []
+    for _ in range(iterations):
         node_labels = weisfeiler_lehman_step(G, node_labels, edge_attr=edge_attr)
-        counter = Counter()
-        # count node labels
-        for node, d in node_labels.items():
-            h = blake2b(digest_size=digest_size)
-            h.update(d.encode("ascii"))
-            counter.update([h.hexdigest()])
+        counter = Counter(node_labels.values())
         # sort the counter, extend total counts
-        items.extend(sorted(counter.items(), key=lambda x: x[0]))
+        subgraph_hash_counts.extend(sorted(counter.items(), key=lambda x: x[0]))
 
     # hash the final counter
-    h = blake2b(digest_size=digest_size)
-    h.update(str(tuple(items)).encode("ascii"))
-    h = h.hexdigest()
-    return h
+    return _hash_label(str(tuple(subgraph_hash_counts)), digest_size)
+
+
+def weisfeiler_lehman_subgraph_hashes(
+    G, edge_attr=None, node_attr=None, iterations=3, digest_size=16
+):
+    """Return a dictionary indexed by node, giving a list of Weisfeiler-Lehman
+    (WL) hashes for each subgraph originating from each node, ordered by
+    iteration depth.
```

Comment:
What do you mean by "subgraph originating from each node"?
Are these induced subgraphs? On what set of nodes?

Also, this should be one-line short description of the function.  Actually, in this case, you should make this first sentence the first paragraph of the doc_string. And make a first line that is short (used to label the links in the doc to this function). Something like:  Return as dict of hashes for each subgraph containing each node.

Reply:
on each iteration, each node label is replaced with the hash of its 1-hop neighbors labels concatenated. So, at iteration k the hash each node label is influenced by all nodes in a radius of 2*k. So, the hash at iteration i for each node n is the WL hash for the subgraph induced by nodes in a radius 2i around n.

It's really hard to put it succinctly but accurately, I feel like english language can handle dicts and lists but plurals start to get confusing when talking about a dict of lists at each iteration! Suggestions welcome!

Reply:
how about:
```
Return a dictionary of subgraph hashes by node. The hashed subgraphs
are induced by increasing k-hop neighborhoods around each node, where
k is even.
```
still too long?

Reply:
You get a hash per iteration, so for each node you get a list in increasing iteration order of hashes.
Maybe the following?
```
Return a dictionary where each key is a node in the graph and each value is a list containing hashes of increasing size subgraphs containing that node. 

The hashed subgraphs are induced on neighbors within k edges of the node, for each even value of k.
```

Reply:
I like the second option because it has a nice short first line the docs can use for links to this function.
The word "even" gets kind of lost in the last sentence. And how do we know when to stop increasing k?
What do you think of:
```
Return a dictionary of subgraph hashes by node. 

The dictionary is keyed by node to a list of hashes in increasingly
sized induced subgraphs containing the nodes within 2*k edges
of the key node for increasing integer k until all nodes are included.
```

Reply:
I think you've cracked it!

Reply:
updated this, hopefully it's clear enough now